libcaf_net/caf/net/multiplexer.cpp

```diff
@@ -53,411 +76,614 @@ const short error_mask = POLLRDHUP | POLLERR | POLLHUP | POLLNVAL;
+private:
+  net::pipe_socket fd_;
+  net::socket_manager* owner_ = nullptr;
+  net::default_multiplexer* mpx_ = nullptr;
+  msg_buf buf_;
+  size_t buf_size_ = 0;
+};
 
-size_t multiplexer::num_socket_managers() const noexcept {
-  return managers_.size();
-}
+/// Multiplexes any number of ::socket_manager objects with a ::socket.
+class default_multiplexer : public detail::atomic_ref_counted,
+                            public multiplexer {
+public:
+  // -- member types -----------------------------------------------------------
 
-ptrdiff_t multiplexer::index_of(const socket_manager_ptr& mgr) const noexcept {
-  auto first = managers_.begin();
-  auto last = managers_.end();
-  auto i = std::find(first, last, mgr);
-  return i == last ? -1 : std::distance(first, i);
-}
+  struct poll_update {
+    short events = 0;
+    socket_manager_ptr mgr;
+  };
 
-ptrdiff_t multiplexer::index_of(socket fd) const noexcept {
-  auto first = pollset_.begin();
-  auto last = pollset_.end();
-  auto i = std::find_if(first, last,
-                        [fd](const pollfd& x) { return x.fd == fd.id; });
-  return i == last ? -1 : std::distance(first, i);
-}
+  using poll_update_map = unordered_flat_map<socket, poll_update>;
 
-middleman& multiplexer::owner() {
-  CAF_ASSERT(owner_ != nullptr);
-  return *owner_;
-}
+  using pollfd_list = std::vector<pollfd>;
 
-actor_system& multiplexer::system() {
-  return owner().system();
-}
+  using manager_list = std::vector<socket_manager_ptr>;
 
-// -- implementation of execution_context --------------------------------------
+  // -- friends ----------------------------------------------------------------
 
-void multiplexer::ref_execution_context() const noexcept {
-  ref();
-}
+  friend class pollset_updater; // Needs access to the `do_*` functions.
 
-void multiplexer::deref_execution_context() const noexcept {
-  deref();
-}
+  // -- constructors, destructors, and assignment operators --------------------
 
-void multiplexer::schedule(action what) {
-  CAF_LOG_TRACE("");
-  if (std::this_thread::get_id() == tid_) {
-    pending_actions_.push_back(what);
-  } else {
-    auto ptr = std::move(what).as_intrusive_ptr().release();
-    write_to_pipe(detail::pollset_updater::code::run_action, ptr);
+  explicit default_multiplexer(middleman* parent) : owner_(parent) {
+    // nop
   }
-}
 
-void multiplexer::watch(disposable what) {
-  watched_.emplace_back(what);
-}
+  // -- implementation of caf::net::multiplexer --------------------------------
+
+  // -- initialization ---------------------------------------------------------
+
+  error init() override {
+    auto pipe_handles = make_pipe();
+    if (!pipe_handles)
+      return std::move(pipe_handles.error());
+    auto updater = pollset_updater::make(pipe_handles->first);
+    auto mgr = socket_manager::make(this, std::move(updater));
+    if (auto err = mgr->start())
+      return err;
```

Comment:
Isn't this leaking the second file descriptor in case of an error?

```suggestion
    if (auto err = mgr->start()) {
      close(pipe_handles->second);
      return err;
    }
```

Reply:
If there's no RAII set up, then yes. I didn't go into detail here.